src/Accounts/Authentication.ResourceManager/AzureRmProfile.cs

```diff
@@ -631,5 +648,133 @@ static string GetJsonText(string text)
 
             return result;
         }
+
+        private void WriteWarning(string message)
+        {
+            EventHandler<StreamEventArgs> writeWarningEvent;
+            if (AzureSession.Instance.TryGetComponent(AzureRMCmdlet.WriteWarningKey, out writeWarningEvent))
+            {
+                writeWarningEvent(this, new StreamEventArgs() { Message = message });
+            }
+        }
+
+        private void RefreshContextsFromCache()
+        {
+            var authenticationClientFactory = new SharedTokenCacheClientFactory();
+            var accounts = authenticationClientFactory.ListAccounts();
+            if (!accounts.Any())
+            {
+                if (!Contexts.Any(c => c.Key != "Default" && c.Value.Account.Type == AzureAccount.AccountType.User))
```

Comment:
I think this method could use a bit of breaking up.

RemoveUserContexts()
RemoveUnmatchedUserContexts()
etc.

Also, as stated above, we need soem tracing here so we can tell why a context was removed if soemthign goes wrong.